File: frontend/src/types/courses.ts
export type CourseStatus = 'draft' | 'published' | 'archived';
export type CourseEnrollmentStatus = 'not_started' | 'learning' | 'completed' | 'cancelled';
export type LessonStatus = 'draft' | 'published' | 'archived';
export type StudentLessonStatus = 'not_started' | 'learning' | 'completed';
export type AssignmentType = 'essay' | 'quiz' | 'practice';
export type SubmissionStatus = 'submitted' | 'ai_graded' | 'teacher_confirmed';

export interface CourseAttachment {
  id: string;
  name: string;
  type: 'video' | 'pdf' | 'document' | 'link';
  url: string;
  size?: string;
}

export interface Lesson {
  id: string;
  course_id: string;
  chapter_title: string;
  title: string;
  description: string;
  content: string;
  duration_mins: number;
  sort_order: number;
  status: LessonStatus;
  attachments: CourseAttachment[];
  student_status?: StudentLessonStatus;
  created_at: string;
  updated_at?: string | null;
}

export interface Course {
  id: string;
  code: string;
  title: string;
  description: string;
  thumbnail: string;
  subject: string;
  grade: string;
  teacher_ids: string[];
  teacher_id: string;
  teacher_name: string;
  assistant_teacher_id?: string;
  assistant_teacher_name?: string;
  goals: string[];
  syllabus_overview: string;
  lesson_count: number;
  assignment_count: number;
  exam_count: number;
  student_count: number;
  start_date: string;
  end_date: string;
  status: CourseStatus;
  created_at: string;
  updated_at?: string | null;
}

export interface CourseEnrollment {
  id: string;
  course_id: string;
  course_code: string;
  course_title: string;
  subject: string;
  grade: string;
  student_id: string;
  student_code: string;
  student_name: string;
  student_email: string;
  teacher_name: string;
  enrollment_date: string;
  status: CourseEnrollmentStatus;
  progress_pct: number;
  gpa_average: number;
  completed_lessons: number;
  total_lessons: number;
  last_activity_at?: string | null;
}

export interface Assignment {
  id: string;
  course_id: string;
  lesson_id?: string | null;
  course_title: string;
  title: string;
  description: string;
  instructions: string;
  assignment_type: AssignmentType;
  due_at: string;
  max_score: number;
  auto_grade: boolean;
  status: 'draft' | 'published' | 'archived';
  submitted_count: number;
  total_students: number;
  created_by: string;
  created_at: string;
  updated_at?: string | null;
}

export interface AIGradingResult {
  score: number;
  feedback: string;
  rubric: Array<Record<string, string | number>>;
}

export interface StudentSubmission {
  id: string;
  assignment_id: string;
  assignment_title: string;
  course_id: string;
  course_title: string;
  student_id: string;
  student_code: string;
  student_name: string;
  submitted_at: string;
  content: string;
  attachment_ids: string[];
  revision_count: number;
  status: 'submitted' | 'ai_grading' | 'ai_suggested' | 'grading_failed' | 'teacher_graded';
  ai_grade?: AIGradingResult | null;
  teacher_score?: number | null;
  teacher_feedback?: string | null;
  graded_by?: string | null;
  graded_at?: string | null;
  final_score?: number | null;
  grading_error?: string | null;
}

export interface AssignmentCreate {
  course_id: string;
  lesson_id?: string | null;
  title: string;
  description?: string;
  instructions?: string;
  assignment_type?: AssignmentType;
  due_at: string;
  max_score: number;
  auto_grade?: boolean;
  status?: Assignment['status'];
}

export type AssignmentUpdate = Partial<Omit<AssignmentCreate, 'course_id'>>;

export interface SubmissionCreate {
  content: string;
  attachment_ids?: string[];
}

export interface TeacherGrade {
  score: number;
  feedback: string;
}

export interface CourseStatSummary {
  total_courses: number;
  active_courses: number;
  total_teachers: number;
  total_students: number;
  total_enrollments: number;
  total_assignments: number;
  total_submissions: number;
  ai_graded_submissions: number;
}

File: backend/app/schemas/courses.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


CourseStatus = Literal["draft", "published", "archived"]
LessonStatus = Literal["draft", "published", "archived"]
EnrollmentStatus = Literal["not_started", "learning", "completed", "cancelled"]


class CourseCreate(BaseModel):
    code: str = Field(min_length=2, max_length=32)
    title: str = Field(min_length=1, max_length=200)
    description: str = Field(default="", max_length=4000)
    thumbnail: str = ""
    subject: str = Field(min_length=1, max_length=120)
    grade: str | None = Field(default=None, max_length=40)
    teacher_ids: list[str] = Field(default_factory=list)
    goals: list[str] = Field(default_factory=list)
    syllabus_overview: str = ""
    start_date: str = ""
    end_date: str = ""
    status: CourseStatus = "draft"


class CourseUpdate(BaseModel):
    code: str | None = Field(default=None, min_length=2, max_length=32)
    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=4000)
    thumbnail: str | None = None
    subject: str | None = Field(default=None, min_length=1, max_length=120)
    grade: str | None = Field(default=None, max_length=40)
    teacher_ids: list[str] | None = None
    goals: list[str] | None = None
    syllabus_overview: str | None = None
    start_date: str | None = None
    end_date: str | None = None
    status: CourseStatus | None = None


class CourseRead(CourseCreate):
    id: str
    teacher_id: str = ""
    teacher_name: str = ""
    lesson_count: int = Field(default=0, ge=0)
    assignment_count: int = Field(default=0, ge=0)
    exam_count: int = Field(default=0, ge=0)
    student_count: int = Field(default=0, ge=0)
    created_at: datetime
    updated_at: datetime | None = None


class CourseAttachment(BaseModel):
    id: str
    name: str
    type: Literal["video", "pdf", "document", "link"]
    url: str
    size: str | None = None


class LessonCreate(BaseModel):
    chapter_title: str = Field(default="", max_length=200)
    title: str = Field(min_length=1, max_length=200)
    description: str = Field(default="", max_length=4000)
    content: str = ""
    duration_mins: int = Field(default=45, ge=0, le=1440)
    sort_order: int | None = Field(default=None, ge=0)
    status: LessonStatus = "draft"
    attachments: list[CourseAttachment] = Field(default_factory=list)


class LessonUpdate(BaseModel):
    chapter_title: str | None = Field(default=None, max_length=200)
    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=4000)
    content: str | None = None
    duration_mins: int | None = Field(default=None, ge=0, le=1440)
    sort_order: int | None = Field(default=None, ge=0)
    status: LessonStatus | None = None
    attachments: list[CourseAttachment] | None = None


class LessonRead(LessonCreate):
    id: str
    course_id: str
    sort_order: int
    created_at: datetime
    updated_at: datetime | None = None


class EnrollmentCreate(BaseModel):
    student_id: str = Field(min_length=1)


class EnrollmentRead(BaseModel):
    id: str
    course_id: str
    course_code: str
    course_title: str
    subject: str
    grade: str = ""
    student_id: str
    student_code: str = ""
    student_name: str = ""
    student_email: str = ""
    teacher_name: str = ""
    enrollment_date: datetime
    status: EnrollmentStatus = "not_started"
    progress_pct: float = Field(default=0, ge=0, le=100)
    gpa_average: float = Field(default=0, ge=0, le=10)
    completed_lessons: int = Field(default=0, ge=0)
    total_lessons: int = Field(default=0, ge=0)
    last_activity_at: datetime | None = None


class CourseStatistics(BaseModel):
    total_courses: int = 0
    active_courses: int = 0
    total_teachers: int = 0
    total_students: int = 0
    total_enrollments: int = 0
    total_assignments: int = 0
    total_submissions: int = 0
    ai_graded_submissions: int = 0
